modules/caddyhttp/replacer.go

```diff
@@ -208,12 +217,79 @@ func addHTTPVarsToReplacer(repl *caddy.Replacer, req *http.Request, w http.Respo
 	repl.Map(httpVars)
 }
 
+func getReqTLSReplacement(req *http.Request, key *string) (string, bool) {
+	if req == nil || req.TLS == nil || key == nil {
+		return "", false
+	}
+
+	if len(*key) < len(reqTLSReplPrefix) {
+		return "", false
+	}
+
+	field := strings.ToLower((*key)[len(reqTLSReplPrefix):])
+
+	if strings.HasPrefix(field, "client.") {
+		cert := getTLSPeerCert(req.TLS)
+		if cert == nil {
+			return "", false
+		}
+
+		switch field {
+		case "client.cert":
+			return string(cert.Raw), true
+		case "client.fingerprint":
+			return fmt.Sprintf("%x", sha256.Sum256(cert.Raw)), true
+		case "client.issuer":
+			return cert.Issuer.String(), true
+		case "client.serial":
+			return fmt.Sprintf("%x", cert.SerialNumber), true
+		case "client.subject":
+			return cert.Subject.String(), true
+		default:
+			return "", false
+		}
+	}
+
+	switch field {
+	case "version":
+		return caddytls.ProtocolName(req.TLS.Version), true
+	case "cipher_suite":
+		return caddytls.CipherSuiteName(req.TLS.CipherSuite), true
+	case "resumed":
+		if req.TLS.DidResume {
+			return "true", true
+		}
+		return "false", true
+	case "proto":
+		return req.TLS.NegotiatedProtocol, true
+	case "proto_mutual":
+		if req.TLS.NegotiatedProtocolIsMutual {
+			return "true", true
+		}
+		return "false", true
+	case "server_name":
+		return req.TLS.ServerName, true
+	default:
+		return "", false
+	}
+}
+
+// getTLSPeerCert retrieves the first peer certificate from a TLS session.
+// Returns nil if no peer cert is in use.
+func getTLSPeerCert(cs *tls.ConnectionState) *x509.Certificate {
+	if len(cs.PeerCertificates) == 0 {
+		return nil
+	}
+	return cs.PeerCertificates[0]
```

Comment:
I can't remember, if the peer presents a certificate chain (bundle), will this be the leaf certificate or the CA? Which one will we want? Does it depend? We should consider these questions.

Reply:
Good question.  I'll try to find an answer.  I'm pretty sure I mirrored the behavior of v1 for this.

Reply:
Here's what I see in the stdlib:

The godocs for [`tls.ConnectionState`](https://golang.org/pkg/crypto/tls/#ConnectionState) `PeerCertificates` are not explicit about the leaf cert being first.  There's a hint in [`tls.Certificate`](https://golang.org/pkg/crypto/tls/#Certificate), which says "A Certificate is a chain of one or more certificates, leaf first," but `PeerCertificates` is not of type `tls.Certificate`.

Browsing through [`crypto/tls/handshake_server.go#L545`](https://github.com/golang/go/blob/88e564edb13f1596c12ad16d5fd3c7ac7deac855/src/crypto/tls/handshake_server.go#L545), it certainly appears that the first certificate is assumed to be the leaf.

Reply:
Indeed, it also seems to be the case here: https://github.com/golang/go/blob/33d20760a23acffd3b200b922c5687ed2aeb74cc/src/crypto/tls/conn.go#L1419

It's not strictly documented anywhere, but let's run with that, I guess.